Realm/RLMQueryUtil.mm

```diff
@@ -625,6 +635,50 @@ void add_collection_operation_constraint(NSPredicateOperatorType operatorType,
     }
 }
 
+// Diacritic-sensitive version.
+void QueryBuilder::add_substring_constraint(const StringData& value, Query&& condition) {
+    // Foundation always returns false for substring operations with a RHS of null or "".
+    m_query.and_query(value.size()
+                      ? std::move(condition)
+                      : std::unique_ptr<Expression>(new FalseExpression));
+}
+
+// Diacritic-sensitive version.
+void QueryBuilder::add_substring_constraint(const Columns<String>& value, Query&& condition) {
+    // Foundation always returns false for substring operations with a RHS of null or "".
+    // We don't need to concern ourselves with the possibility of column2 traversing a link list
+    // and producing multiple values per row as such expressions will have been rejected.
+    m_query.group();
+    m_query.and_query(value != null() && value != "");
+    m_query.and_query(std::move(condition));
```

Comment:
Hmm, we should fix `size()` to be const since there's no reason I can see for it to be non-const. I'd suggest doing that, and working around that for now using a `const_cast`.